### PR TITLE
[BE] deprecate llama4, move apply_fsdp to common file

### PR DESCRIPTION
**Pull request by tianyu-l**
Three major changes caused by removal of `llama4` folder
- llama3 defines `apply_fsdp` shared by dense models, llama4 defines another `apply_fsdp` shared by MoE models which is a superset of dense `apply_fsdp`. Now consolidated to `apply_fsdp_to_decoder` in distributed/fsdp.py
- llama4 uses more than one flex attention masks (RoPE vs NoPE). This is also the case for gpt-oss https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/gpt_oss/model.py#L256, so we still have that showcased.
- Some llama4 layers don't have RoPE module (NoPE layers). This PR removes this optionality because all other decoder models have RoPE in every layer.

### CODE REVIEW

File: torchtitan/distributed/fsdp.py [hunk 1]
```diff
@@ -60,3 +73,264 @@ def get_fsdp_reshard_after_forward_policy(
...
+    FSDP unit and the expert-parallel prefetching below is skipped.
+
+    Args:
+        model (nn.Module): The model to apply data parallelism to.
+        dp_mesh (DeviceMesh): The device mesh to use for data parallelism.
+        param_dtype (torch.dtype): The data type to use for model parameters.
+        reduce_dtype (torch.dtype): The data type to use for reduction operations.
+        pp_enabled (bool): Whether pipeline parallelism is enabled.
+        cpu_offload (bool, optional): Whether to offload model parameters to CPU. Defaults to False.
+        reshard_after_forward_policy (str, optional): The policy to use for resharding after forward pass. Defaults to "default".
```
Comment (fegin) [inline torchtitan/distributed/fsdp.py]: nit: this line is way too long, consider to ask claude to cap the docstring maximum characters

File: torchtitan/distributed/fsdp.py [hunk 2]
```diff
@@ -60,3 +73,264 @@ def get_fsdp_reshard_after_forward_policy(
...
+            Other options: "never", "always".
+            - "default" applies default resharding behavior, implementing "smart defaults" for known optimal scenarios.
```
Comment (fegin) [inline torchtitan/distributed/fsdp.py]: ditto

File: torchtitan/distributed/fsdp.py [hunk 3]
```diff
@@ -60,3 +73,264 @@ def get_fsdp_reshard_after_forward_policy(
...
+        logger.info("Applied FSDP to the model")
+    if cpu_offload:
+        logger.info("Applied CPU Offloading to the model")
+
+    # NOTE: set up explicit prefetching when EP is enabled, as D2H syncs
+    # in EP could interfere with implicit prefetching in FSDP
+    if ep_degree == 1:
+        return
+
+    # forward
```
Comment (fegin) [inline torchtitan/distributed/fsdp.py]: this comment doesn't give too much information. We can reuse some wordings from line 291: "set up explicit prefetching when EP is enabled for forward".

File: torchtitan/distributed/fsdp.py [hunk 4]
```diff
@@ -60,3 +73,264 @@ def get_fsdp_reshard_after_forward_policy(
             raise ValueError(
                 f"Invalid reshard_after_forward_policy: {reshard_after_forward_policy}."
             )
+
+
+def apply_fsdp_to_decoder(
+    model: nn.Module,
```
Comment (fegin) [inline torchtitan/distributed/fsdp.py]: I guess if you type as Decoder, there should be less pyre-ignore below. Or you can use `assert isinstance(model, Decoder)`, which can cause less problems on the caller side.

File: torchtitan/distributed/fsdp.py [hunk 5]
```diff
@@ -60,3 +73,264 @@ def get_fsdp_reshard_after_forward_policy(
...
+    if dp_mesh_dims is not None:
+        fsdp_config["dp_mesh_dims"] = dp_mesh_dims
+    if cpu_offload:
+        fsdp_config["offload_policy"] = CPUOffloadPolicy()
+
+    reshard_after_forward = get_fsdp_reshard_after_forward_policy(
+        reshard_after_forward_policy, pp_enabled
+    )
+
+    if getattr(model, "enable_weight_tying", False):
```
Comment (fegin) [inline torchtitan/distributed/fsdp.py]: Would it be better to pass this as an argument to `apply_fsdp_to_decoder`? A model should know whether `enable_weight_tying` is an option to that model or not.

File: torchtitan/distributed/fsdp.py [hunk 6]
```diff
@@ -60,3 +73,264 @@ def get_fsdp_reshard_after_forward_policy(
...
+        if next_transformer_block is not None:
+            # pyrefly: ignore [missing-attribute]
+            transformer_block.set_modules_to_forward_prefetch([next_transformer_block])
+        elif model.norm is not None and model.lm_head is not None:
+            # pyrefly: ignore [missing-attribute]
+            transformer_block.set_modules_to_forward_prefetch(
+                [model.norm, model.lm_head]
+            )
+
+    # backward
```
Comment (fegin) [inline torchtitan/distributed/fsdp.py]: ditto

**Review by fegin — APPROVED** [review]
LGTM, most comments are about nits.

File: torchtitan/distributed/fsdp.py (hunk 5 above)
Comment (tianyu-l) [reply torchtitan/distributed/fsdp.py]: I think `enable_weight_tying` is a general `Decoder` concept (used by all small models), so I moved it into `Decoder`.

**Review by SherlockNoMad — APPROVED** [review]
